File: 2021/deno/day-3.ts
const input = await Deno.readTextFile("./input/day-3.in");

const rows = input.split("\n");

const binaries = rows.map((row) => row.split(""));

const len = binaries[0].length;

const gamma = Array.from({ length: len }, (_, i) => {
  const ones = binaries.reduce(
    (acc, bin) => (bin[i] === "1" ? acc + 1 : acc),
    0
  );
  const zeros = binaries.reduce(
    (acc, bin) => (bin[i] === "0" ? acc + 1 : acc),
    0
  );

  return ones > zeros ? "1" : "0";
}).join("");

const epsilon = Array.from({ length: len }, (_, i) => {
  const ones = binaries.reduce(
    (acc, bin) => (bin[i] === "1" ? acc + 1 : acc),
    0
  );
  const zeros = binaries.reduce(
    (acc, bin) => (bin[i] === "0" ? acc + 1 : acc),
    0
  );

  return zeros > ones ? "1" : "0";
}).join("");

/**
 * Part One
 */

console.log("Part One:", parseInt(gamma, 2) * parseInt(epsilon, 2));

/**
 * Part Two
 */

let bit = 0; // first bit
let oxygen = binaries.slice(0);
let co2 = binaries.slice(0);

while (1) {
  if (oxygen.length === 1) {
    break;
  }
  const ones = oxygen.reduce(
    (acc, bin) => (bin[bit] === "1" ? acc + 1 : acc),
    0
  );
  const zeros = oxygen.reduce(
    (acc, bin) => (bin[bit] === "0" ? acc + 1 : acc),
    0
  );

  if (ones >= zeros) {
    oxygen = oxygen.filter((bin) => bin[bit] === "1");
  } else {
    oxygen = oxygen.filter((bin) => bin[bit] === "0");
  }

  bit = (bit + 1) % len;
}

// reset the bit
bit = 0;

while (1) {
  if (co2.length === 1) {
    break;
  }
  const ones = co2.reduce((acc, bin) => (bin[bit] === "1" ? acc + 1 : acc), 0);
  const zeros = co2.reduce((acc, bin) => (bin[bit] === "0" ? acc + 1 : acc), 0);

  if (zeros > ones) {
    co2 = co2.filter((bin) => bin[bit] === "1");
  } else {
    co2 = co2.filter((bin) => bin[bit] === "0");
  }

  bit = (bit + 1) % len;
}

console.log(
  "Part Two:",
  parseInt(oxygen[0].join(""), 2) * parseInt(co2[0].join(""), 2)
);

File: 2021/swift/day-3.swift
import Foundation

let filename = "../input/day-3.in"

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let rows = contents.split(separator: "\n").map { String($0) }

        var cols = [String](repeating: "", count: rows[0].count)

        for row in rows {
            for (index, bit) in row.enumerated() {
                cols[index] += String(bit)
            }
        }

        var gamma = ""
        var epsilon = ""

        for col in cols {
            let ones = Array(col).filter { $0 == "1" }.count > (col.count / 2)

            // calculating epsilon as 2-complemented would blow cause an overflow
            gamma += ones ? "1" : "0"
            epsilon += ones ? "0" : "1"
        }

        print("Part one:", Int(gamma, radix: 2)! * Int(epsilon, radix: 2)!)

        let ratingCalculator: (String, (Int, Int) -> Character) -> String = {
            var candidates = $0.split(separator: "\n").map { Array(String($0)) }

            let upperBound = candidates[0].count

            var index = 0

            while candidates.count > 1 {
                let ones = candidates.filter {
                    $0[index] == "1"
                }.count

                let predicate: Character = $1(candidates.count - ones, ones)

                candidates = candidates.filter { $0[index] == predicate }

                index = index + 1 % upperBound
            }

            return candidates[0].map(String.init).reduce("", +)
        }

        let oxygenRating = Int(ratingCalculator(contents) { $1 >= $0 ? "1" : "0" }, radix: 2)!
        let co2Rating = Int(ratingCalculator(contents) { $1 >= $0 ? "0" : "1" }, radix: 2)!

        print("Part two:", oxygenRating * co2Rating)

    } catch {
        print("Error reading file")
    }
}

main()
